crates/relayer/relayer-components/src/chain/traits/queries/client_status.rs

```diff
@@ -0,0 +1,99 @@
+use cgp_core::prelude::*;
+
+use crate::chain::traits::types::client_state::{HasClientStateFields, HasClientStateType};
+use crate::chain::traits::types::consensus_state::{
+    HasConsensusStateFields, HasConsensusStateType,
+};
+use crate::chain::traits::types::ibc::HasIbcChainTypes;
+
+use super::chain_status::CanQueryChainStatus;
+use super::client_state::{CanQueryClientState, CanQueryClientStateWithHeight};
+use super::consensus_state::CanQueryConsensusState;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ClientStatus {
+    Frozen,
+    Expired,
+    Active,
+}
+
+#[async_trait]
+pub trait CanQueryClientStatus<Counterparty>:
+    HasIbcChainTypes<Counterparty>
+    + HasErrorType
+    + CanQueryClientState<Counterparty>
+    + CanQueryClientStateWithHeight<Counterparty>
+    + CanQueryChainStatus
+    + CanQueryConsensusState<Counterparty>
+where
+    Counterparty: HasIbcChainTypes<Self>
+        + HasClientStateType<Self>
+        + HasClientStateFields<Self>
+        + HasConsensusStateType<Self>
+        + HasConsensusStateFields<Self>,
+{
+    async fn query_client_status(
+        &self,
+        client_id: &Self::ClientId,
+    ) -> Result<ClientStatus, Self::Error>;
+}
+
+#[async_trait]
+impl<Chain, Counterparty> CanQueryClientStatus<Counterparty> for Chain
+where
+    Chain: HasIbcChainTypes<Counterparty>
+        + HasErrorType
+        + CanQueryClientState<Counterparty>
+        + CanQueryClientStateWithHeight<Counterparty>
+        + CanQueryChainStatus
+        + CanQueryConsensusState<Counterparty>,
+    Counterparty: HasIbcChainTypes<Chain>
+        + HasClientStateType<Chain>
+        + HasClientStateFields<Chain>
+        + HasConsensusStateType<Chain>
+        + HasConsensusStateFields<Chain>,
+{
+    async fn query_client_status(
+        &self,
+        client_id: &Self::ClientId,
+    ) -> Result<ClientStatus, Self::Error> {
+        let client_state = self.query_client_state(client_id).await?;
+        // .wrap_err_with(|e| "Failed to query client state for client `{client_id}`")?;
```

Comment:
What's the best way to add additional context to errors yield in here?

Reply:
I thought about this before, and I am thinking on experimenting with using the pattern `CanRaiseError<CustomError>` to add details for various errors.

For this case here, we would define an error struct `QueryClientStateError` as follows:

```rust
pub struct QueryClientStateError<'a, Chain, Counterparty>
where
    Chain: HasIbcChainTypes<Counterparty> + HasErrorType,
{
    pub error: Chain::Error,
    pub client_id: &'a Chain::ClientId,
}
```

We then add the trait bound `for<'a> CanRaiseError<QueryClientStateError<'a, Chain, Counterparty>>` to require the chain context to handle the case for client state error.

This would require `CosmosChain` to implement the particular error handler, which is handled by `HandleCosmosError`. So we would have an implementation like:

```rust
impl<'a, Context, Chain, Counterparty>
    ErrorRaiser<Context, QueryClientStateError<'a, Chain, Counterparty>> for HandleCosmosError
where
    Context: HasErrorType<Error = Error>,
    Chain: HasIbcChainTypes<Counterparty> + HasErrorType,
{
    fn raise_error(e: QueryClientStateError<'a, Chain, Counterparty>) -> Error {
        BaseError::generic(eyre!(
            "Failed to query client state for client `{}`: {:?}",
            e.client_id,
            e.error,
        ))
        .into()
    }
}
```

With this, the implementation would be given the flexibility to customize on how it wants to store the error. We could create a new error variant in the concrete error type, or we could just return the inner error and discard the extra details.

One potential tradeoff for this approach is that as the number of error types grow, the concrete context would have to handle every case even if they are not interested with the extra detail. A way to ease default error handling is by implementing traits like `Debug` for the error detail type:

```rust
impl<'a, Chain, Counterparty> Debug for QueryClientStateError<'a, Chain, Counterparty>
where
    Chain: HasIbcChainTypes<Counterparty> + HasErrorType,
    Chain::ClientId: Debug,
{
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        writeln!(
            f,
            "Failed to query client state for client `{}`",
            self.client_id
        )?;
        writeln!(f, "{:?}", self.error)?;
        Ok(())
    }
}
```

This would allow catch all error raiser implementations like `RaiseDebugError` from `cgp-error-eyre` to handle the errors generically.

This pattern may result in quite a bit more code, but I think it helps give the flexibility of error handling to be as simple or as complex as the concrete error implementation needs.

Reply:
Basically with this pattern, we will be able to implement catch-all error wrapping handlers like `impl<Detail> ErrorRaiser<(Detail, Error)> for ErrorHandler`.

Reply:
Thanks for the detailed write-up, let's do this in a follow-up PR then.

Reply:
Tracked in https://github.com/informalsystems/hermes-sdk/issues/168